sbom/spdx/encoder.go

```diff
@@ -0,0 +1,301 @@
+// We first convert the IndexReport to an SPDX doc of the latest version, then
+// convert that doc to the specified version. We assume there's no data munging
+// going from latest to the specified version.
+func (e *Encoder) Encode(ctx context.Context, ir *claircore.IndexReport) (io.Reader, error) {
+	spdx, err := e.parseIndexReport(ctx, ir)
+	if err != nil {
+		return nil, err
+	}
+
+	var tmpConverterDoc common.AnyDocument
+	switch e.Version {
+	case V2_1:
+		var targetDoc v2_1.Document
+		if err := convert.Document(spdx, targetDoc); err != nil {
+			return nil, err
+		}
+		tmpConverterDoc = targetDoc
+	case V2_2:
+		var targetDoc v2_2.Document
+		if err := convert.Document(spdx, targetDoc); err != nil {
+			return nil, err
+		}
+		tmpConverterDoc = targetDoc
+	case V2_3:
+		// parseIndexReport currently returns a v2_3.Document so do nothing
+		tmpConverterDoc = spdx
+	default:
+		return nil, fmt.Errorf("unknown SPDX version: %v", e.Version)
+	}
+
+	switch e.Format {
+	case JSON:
+		// TODO(DO NOT MERGE): Should this be outside the switch? i.e., should
+		//  we use this for all formatting cases? We could return it outside the
+		//  switch as the "default" case
+		buf := &bytes.Buffer{}
+		if err := spdxjson.Write(tmpConverterDoc, buf); err != nil {
+			return nil, err
+		}
+		return buf, nil
+	}
+
+	return nil, fmt.Errorf("unknown requested format: %v", e.Format)
+}
+
+func (e *Encoder) parseIndexReport(ctx context.Context, ir *claircore.IndexReport) (*v2_3.Document, error) {
+	creatorInfo := e.Creators
+	spdxCreators := make([]v2common.Creator, len(creatorInfo))
+	for i, creator := range creatorInfo {
+		spdxCreators[i].Creator = creator.Creator
+		spdxCreators[i].CreatorType = creator.CreatorType
+	}
+
+	// Initial metadata
+	out := &v2_3.Document{
+		SPDXVersion:    v2_3.Version,
+		DataLicense:    v2_3.DataLicense,
+		SPDXIdentifier: "DOCUMENT",
+		// TODO(DO NOT MERGE): Is this ok?
+		DocumentName:      ir.Hash.String(),
+		DocumentNamespace: e.DocumentNamespace,
+		CreationInfo: &v2_3.CreationInfo{
+			Creators: spdxCreators,
+			Created:  time.Now().Format("2006-01-02T15:04:05Z"),
+		},
+		DocumentComment: e.DocumentComment,
+	}
+
+	var rels []*v2_3.Relationship
+	repoMap := map[string]*v2_3.Package{}
+	distMap := map[string]*v2_3.Package{}
+	pkgMap := map[string]*v2_3.Package{}
+	for _, r := range ir.IndexRecords() {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		// This could happen if the PackageScanner that found this package is
+		// associated with two different Ecosystems and one of those Ecosystems
+		// doesn't have the RepositoryScanner. If something like that happens,
+		// we'll have the Repository information in another IndexRecord.
+		if r.Repository == nil || r.Repository.ID == "" {
+			continue
+		}
+
+		pkg, ok := pkgMap[r.Package.ID]
+
+		// TODO(DO NOT MERGE): Is there a way to distinguish "APPLICATION" and "LIBRARY" packages? Probably not but
```

Comment:
IMO if we are talking about distinguishing between language packages and package manager packages, then it's moot because we can know definitively from the index report IIRC

Reply:
Sorry, forgot the remove this comment after we discussed it the other day.